test-framework/jacoco/runtime/src/test/java/io/quarkus/jacoco/runtime/DataFileWatchTest.java

```diff
@@ -0,0 +1,229 @@
+package io.quarkus.jacoco.runtime;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+public class DataFileWatchTest {
+
+    @TempDir
+    Path tempDir;
+
+    @Test
+    public void dataFileSize() throws Exception {
+        Path datafile = tempDir.resolve("some-data-file.exec");
+
+        DataFileWatch dataFileWatch = new DataFileWatch(datafile, () -> true);
+
+        assertThat(dataFileWatch.dataFileSize()).isEqualTo(-1L);
+
+        Files.createFile(datafile);
+        assertThat(dataFileWatch.dataFileSize()).isEqualTo(0L);
+
+        Files.write(datafile, new byte[100]);
+        assertThat(dataFileWatch.dataFileSize()).isEqualTo(100L);
+    }
+
+    @Test
+    public void waitForDataFileGoodCase() throws Exception {
+        // Scenario: Jacoco data file does not already exist, file created after 2 wait iterations, Jacoco
+        // shutdown hook finishes after 3 wait iterations.
+
+        Path datafile = tempDir.resolve("some-data-file.exec");
+
+        AtomicBoolean jacocoFinished = new AtomicBoolean(false);
+        AtomicLong clock = new AtomicLong();
+        AtomicInteger waitIterations = new AtomicInteger(5);
+
+        DataFileWatch dataFileWatch = new DataFileWatch(clock::get, datafile, jacocoFinished::get) {
+            @Override
+            void waitSleep() {
+                switch (waitIterations.decrementAndGet()) {
```

Comment:
Just a suggestion: Use a `CyclicBarrier` or a `Semaphore` to wait/flip between the watch thread and the test thread and control side effects from the test thread in a simple top-down sequential code block.